src/omv/ports/stm32/omv_gpu.c

```diff
@@ -0,0 +1,223 @@
+        cfg.pCLUT = clut;
+        cfg.CLUTColorMode = DMA2D_CCM_ARGB8888;
+        cfg.Size = 255;
+        #if __DCACHE_PRESENT
+        SCB_CleanDCache_by_Addr(clut, 256 * sizeof(uint32_t));
+        #endif
+        HAL_DMA2D_CLUTLoad(&dma2d, cfg, 1);
+        HAL_DMA2D_PollForTransfer(&dma2d, 1000);
+    } else {
+        dma2d.LayerCfg[1].InputColorMode = DMA2D_INPUT_RGB565;
+        dma2d.LayerCfg[1].AlphaMode = DMA2D_REPLACE_ALPHA;
+    }
+
+    dma2d.LayerCfg[1].InputOffset = data->src_img->w - src_width;
+    dma2d.LayerCfg[1].InputAlpha = fast_roundf((data->alpha * 255) / 256.0f);
+    #if defined(MCU_SERIES_F7) || defined(MCU_SERIES_H7)
+    dma2d.LayerCfg[1].AlphaInverted = DMA2D_REGULAR_ALPHA;
+    dma2d.LayerCfg[1].RedBlueSwap = DMA2D_RB_REGULAR;
+    #endif
+    #if defined(MCU_SERIES_H7)
+    dma2d.LayerCfg[1].ChromaSubSampling = DMA2D_NO_CSS;
+    #endif
+
+    HAL_DMA2D_ConfigLayer(&dma2d, 1);
+
+    uint16_t *dst16 = IMAGE_COMPUTE_RGB565_PIXEL_ROW_PTR(data->dst_img, data->dst_y_start) + data->dst_x_start;
+
+    #if __DCACHE_PRESENT
+    // Ensures any cached writes to dst16 are flushed.
+    uint16_t *dst16_tmp = dst16;
+    for (int i = 0; i < dst_height; i++) {
+        SCB_CleanInvalidateDCache_by_Addr(dst16_tmp, dst_width * sizeof(uint16_t));
+        dst16_tmp += data->dst_img->w;
+    }
+    #endif
+
+    uint32_t src;
+
+    if (data->src_img->pixfmt == PIXFORMAT_GRAYSCALE) {
+        uint8_t *src8 = IMAGE_COMPUTE_GRAYSCALE_PIXEL_ROW_PTR(data->src_img, data->src_y_start) + data->src_x_start;
+        src = (uint32_t) src8;
+
+        #if __DCACHE_PRESENT
+        uint8_t *src8_tmp = src8;
+        for (int i = 0; i < src_height; i++) {
+            SCB_CleanDCache_by_Addr(src8_tmp, src_width);
+            src8_tmp += data->src_img->w;
+        }
+        #endif
+    } else {
+        uint16_t *src16 = IMAGE_COMPUTE_RGB565_PIXEL_ROW_PTR(data->src_img, data->src_y_start) + data->src_x_start;
+        src = (uint32_t) src16;
+
+        #if __DCACHE_PRESENT
+        uint16_t *src16_tmp = src16;
+        for (int i = 0; i < src_height; i++) {
+            SCB_CleanDCache_by_Addr(src16_tmp, src_width * sizeof(uint16_t));
+            src16_tmp += data->src_img->w;
+        }
+        #endif
+    }
+
+    HAL_DMA2D_BlendingStart(&dma2d, src, (uint32_t) dst16, (uint32_t) dst16, dst_width, dst_height);
+    HAL_DMA2D_PollForTransfer(&dma2d, 1000);
+
+    #if __DCACHE_PRESENT
+    // Ensures any cached reads to dst16 are dropped.
+    dst16_tmp = dst16;
+    for (int i = 0; i < dst_height; i++) {
+        SCB_InvalidateDCache_by_Addr(dst16_tmp, dst_width * sizeof(uint16_t));
+        dst16_tmp += data->dst_img->w;
+    }
+    #endif
+
+    HAL_DMA2D_DeInit(&dma2d);
+
+    if (data->src_img->pixfmt == PIXFORMAT_GRAYSCALE) {
+        fb_free(); // clut...
+    }
+
+    #if (TIME_GPU == 1)
+    printf("draw time: %u ms\n", mp_hal_ticks_ms() - start);
+    #endif
+
+    return true;
+}
+
+void imlib_hardware_gpu_init() {
```

Comment:
You could save some time if you init/deinit once on soft-reboot loop (if possible). Other than that, these should be.

```C
omv_gpu_init() {
// init should be called on soft-reboot
}

omv_gpu_deinit() {
// deinit should be called on soft-reboot
}

omv_gpu_draw_image() {
}
```

Reply:
Done